File: src/sonar.cpp
#include <iostream>
#include <wiringPi.h>


using namespace std;

int main(void){
        wiringPiSetup();
        unsigned int timeout = 20000;
        int pin = 7;
        while(1){
		unsigned long int starttime, endtime;
                pinMode(pin, OUTPUT);
                digitalWrite(pin,0);
                delayMicroseconds(2);
                digitalWrite(pin, 1);
                delayMicroseconds(5);
                digitalWrite(pin,0);
                pinMode(pin, INPUT);
                bool goodread=true;
                unsigned long int watchtime=micros();
                while(digitalRead(pin)==0 && goodread){
                        starttime=micros();
                        if(starttime-watchtime > timeout){
                                goodread=false;
	                }
                }
                if(goodread){
                        watchtime=micros();
                        while(digitalRead(pin) && goodread){
                                endtime=micros();
                                if(endtime-watchtime>timeout){
                                        goodread=false;
                                }
                        }
                }

                if(goodread){
                        unsigned int duration=endtime-starttime;
                        float distance=(duration*34.0)/2000.0;
                        cout << distance << endl;

                }


        }
	return 0;
}

File: script/sonar.py
import time
import RPi.GPIO as GPIO

GPIO.setmode(GPIO.BOARD)
GPIO.setwarnings(False)
timeout = 0.020
spin=7
while 1:
        GPIO.setup(spin, GPIO.OUT)
        #cleanup output
        GPIO.output(spin, 0)

        time.sleep(0.000002)

        #send signal
        GPIO.output(spin, 1)

        time.sleep(0.000005)

        GPIO.output(spin, 0)

        GPIO.setup(spin, GPIO.IN)
        
        goodread=True
        watchtime=time.time()
        while GPIO.input(spin)==0 and goodread:
                starttime=time.time()
                if (starttime-watchtime > timeout):
                        goodread=False

        if goodread:
                watchtime=time.time()
                while GPIO.input(spin)==1 and goodread:
                        endtime=time.time()
                        if (endtime-watchtime > timeout):
                                goodread=False
        
        if goodread:
                duration=endtime-starttime
                distance=duration*34000/2
                print distance  